### DISCUSSION TITLE
Where to configure relations for self-hosted keto?

### DISCUSSION THREAD

**Discussion by WoodyWoodsta** [ory/keto#1450] [Q&A] [answered]
I've got a self-hosted Ory stack set up in kubernetes.

I've spent a day or so trying to figure out how to set up a simple subject-group-object access control ruleset. I'm now clear on subject-sets, to be used for implicit access to objects via groups, but what is very unclear to me is how to configure relations.

After stringing together various bits of documentation between ory network and ory self-hosted sections of the site, as well as PRs and issues, it seems like there are 3 ways to setup the relations:
1. Via an OPL file (written in TS??)
2. Via the APIs (HTTP/CLI etc)
3. Via the config file, under `namespaces`

For 1, where should this OPL file be and does it completely replace any namespaces specified in the config file?

For 3, a few PRs and issues mention providing a config for each namespace, which I assume would have the relation tuples in that, but in the json schema, the `config` object just has "to be defined" and I don't seem to be able to find any documentation around that.

What is the recommended way for setting up the graph?

Otherwise, thank you for a great set of tools. Once you get around the learning curve of how each component works, it's great software!

**Comment by dankobg** [comment, marked as answer]
Apparently in keto.yaml you can specify

```
namespaces:
  location: file:///etc/config/keto/opl_dir
```
It will create those namespaces from opl file, not sure if it should do anything else since there is zero documentation.

**Comment by rrmistry** [comment]
Reference to keto config schema where this is kind-of documented

> https://github.com/ory/keto/blob/a67586fb9aaf33f7ae4985caea39d65e152117ba/.schema/config.schema.json#L337-L357

Though agree that it should have better documentation in [ory/docs](https://github.com/ory/docs)
